src/query/src/range_select/plan.rs

```diff
@@ -117,32 +120,88 @@ impl Fill {
                             data[i] = data[i - 1].clone()
                         }
                     }
-                    Fill::Linear => {
-                        if 0 < i && i < len - 1 {
-                            match (&data[i - 1], &data[i + 1]) {
-                                (ScalarValue::Float64(Some(a)), ScalarValue::Float64(Some(b))) => {
-                                    data[i] = ScalarValue::Float64(Some((a + b) / 2.0));
-                                }
-                                (ScalarValue::Float32(Some(a)), ScalarValue::Float32(Some(b))) => {
-                                    data[i] = ScalarValue::Float32(Some((a + b) / 2.0));
-                                }
-                                (a, b) => {
-                                    if !a.is_null() && !b.is_null() {
-                                        return Err(DataFusionError::Execution(
-                                            "RangePlan: Apply Fill LINEAR strategy on Non-floating type".to_string()));
-                                    } else {
-                                        continue;
-                                    }
-                                }
-                            }
-                        }
-                    }
+                    Fill::Linear => {}
                     Fill::Const(v) => data[i] = v.clone(),
                 }
             }
         }
         Ok(())
     }
+
+    fn fill_linear(ts: &[i64], data: &mut [ScalarValue]) -> DfResult<()> {
+        let not_null_num = data
+            .iter()
+            .fold(0, |acc, x| if x.is_null() { acc } else { acc + 1 });
+        // We need at least two non-empty data points to perform linear interpolation
+        if not_null_num < 2 {
+            return Ok(());
+        }
+        let mut index = 0;
+        while index < data.len() {
+            // find null interval [start, end)
+            let start = data[index..]
+                .iter()
+                .position(ScalarValue::is_null)
```

Comment:
Looks like it's not efficient. I think we can find all the null indexes before this iteration. And use this info to replace the `start` and `end` calculation.

Reply:
Even if you find it out in advance, you still need such iteration, and pay extra memory overhead to store `start`/`end`. The whole implementation is `O(n)`. Each scan starts from the next interval where an interval may occur (from `index`, so no duplicate iteration), but we can indeed avoid secondary traversal by storing the head and tail null intervals.

Reply:
> we can indeed avoid secondary traversal by storing the head and tail null intervals.

That would be great!